### PR TITLE
#184637150 product review

### PR DESCRIPTION
**Pull request by JENNY-257**
What does this PR do?

     - A buyer should be able to make review on product including rating and feedback

Description of Task to be completed?

      - When a buyer has bought the product  must be able  to make a review  on that product has purchased.

 
How should this be manually tested?

        - Clone the repository
        - Change the directory to destructors-ec-bn by running cd destructors-ec-bn in you terminal
        - Checkout ft-product-review-184637150 branch by running git checkout ft-product-review-184637150 in you terminal
        - open .env.example to check how you can create you .env file
        - Create your .env file
        - Then install dependencies and devdependencies used by running npm install in you terminal
        - To create the models used in this project run npm migrate
        - Run the server by running npm run dev in your terminal
           open you browser and in URL field type http://localhost:PORT/api-docs (the PORT is the port you have written in your .env 
             file. The default port is 3000)
        - Login users  as buyer using /api/v1/users/login endpoint
        - Review product by using  this endpoint  /api/v1/products/:id/reviews

Any background context you want to provide?
           None

What are the relevant pivotal tracker/Trello stories?
           tracker: #184637150

### CODE REVIEW

**Comment by ba-lambert** [comment]
check your test builc

File: src/controllers/productReview.controller.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  const { id: productId } = req.params;
+  const buyerId = req.user.id;
+  const { rating, feedback } = req.body;
+
+  if (req.user.role !== 'buyer') {
+    return res.status(401).json({ message: 'Unauthorized' });
+  }
+
+  try {
+    const product = await Products.findOne({ where: { id: productId } });
```
Comment (ba-lambert) [inline src/controllers/productReview.controller.js]: For best practice this should not be in controller should be in middleware or services

 thanks @ba-lambert  , I have fixed comments

File: src/controllers/productReview.controller.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+    if (!product) {
+      return res.status(404).json({ error: 'Product not found' });
+    }
+    const existingReview = await Reviews.findOne({
```
Comment (ba-lambert) [inline src/controllers/productReview.controller.js:14]: Also this

File: src/database/migrations/20230316084558-create_products_table.js [hunk 3]
```diff
@@ -51,6 +53,9 @@ module.exports = {
           key: 'id',
         },
       },
+      averageRating: {
```
Comment (ba-lambert) [inline src/database/migrations/20230316084558-create_products_table.js]: Here I think you should add migration modification not destroy hard code like you did  here @JENNY-257

File: src/controllers/productReview.controller.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+      });
+  
+      if (existingReview) {
+        return res.status(400).json({ message: 'You have already reviewed this product' });
+      }
+    const review = await Reviews.create({ buyerId, productId, rating, feedback });
+
+    const reviews = await Reviews.findAll({ where: { productId } });
+
+    const totalRatings = reviews.reduce((acc, cur) => acc + cur.rating, 0);
```
Comment (ba-lambert) [inline src/controllers/productReview.controller.js]: can you try to use middleware @JENNY-257

File: __test__/product.test.js [hunk 5]
```diff
@@ -470,3 +471,64 @@ describe('Search products endpoint', () => {
...
+      .post(`/api/v1/products/${product.id}/reviews`)
+      .set('Authorization', `Bearer ${sellerToken}`)
+      .send({rating:3, feedback:'This product is great!'});
+    expect(response.status).toBe(401);
+  });
+  test('should return 404 status code if product to review is not found', async () => {
+    const user = await User.findOne({ where: { email: 'testbuyer@gmail.com' } });
+    buyerToken = generateToken(user);
+    const res = await request(app)
+      .post('/api/v1/products/9974076f-e16a-486f-a923-362ec1747a12/reviews')
```
Comment (Abdulkeza) [inline __test__/product.test.js:525]: Can we stop hard coding IDs like this, please?

means you can use the one you created above.

File: __test__/product.test.js [hunk 6]
```diff
@@ -7,6 +7,7 @@ import { Products } from '../src/database/models';
 import { Categories } from '../src/database/models';
 import { generateToken } from '../src/utils/generateToken';
 import verfyToken from '../src/utils/verifytoken';
+import { Reviews } from '../src/database/models';
```
Comment (Abdulkeza) [inline __test__/product.test.js]: if we don't need this, please can we remove them here?

File: src/ validations/productReview.validation.js [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+    const { error } = reviewProductSchema.validate(req.body ,{
+        abortEarly: false,
+      });
+      if (error) {
+        return res.status(400).json({
+          error: error.details[0].message.replace(/[^a-zA-Z0-9 ]/g, ''),
+        });
+      }
+    next();
+  };
```
Comment (Abdulkeza) [inline src/ validations/productReview.validation.js:16]: Leave one empty line at the end / Bottom of the file

File: src/controllers/productReview.controller.js (hunk 1 above)
Comment (Abdulkeza) [reply src/controllers/productReview.controller.js]: Great @ba-lambert , 

@JENNY-257  we have function called **findProductById** under _product.service.js_ 
let's make use of the function that's already in place.


good thanks @Abdulkeza 
I'm done  to work on  comments

**Review by Abdulkeza — CHANGES_REQUESTED** [review]
@JENNY-257  you did a great JOB,

try to solve the feedback below and you are good to go  :fire:

**Review by ba-lambert — APPROVED** [review]

**Review by nkurunziza1 — APPROVED** [review]
### ### ****

**Pull request merged**